### PR TITLE
Fix: lis3mdl handling when multiple sensors are available

### LINKED ISSUE #10740: lis3mdl with two sensors not working
I found a problem with the starting order if two lis3mdl sensors are available.

Version: stable branch

In QGroundControl PREFLIGHT FAIL: MAG 1 SELFTEST FAILED showed up, so I tested some things.

Configuration: Pixracer r15 with lis3mdl onboard and lis3mdl on GPS module:
https://store.mrobotics.io/mRo-GPS-u-Blox-Neo-M8N-HMC5983-Compass-p/mro-gps003-mr.htm

In the default rc.sensors for PX4FMU_V4, the order is the following:
lis3mdl -X start
lis3mdl start

When I display the sensor values, the onboard sensor shows all 0:

```
nsh> lis3mdl stop
nsh> lis3mdl -X start
LIS3MDL_I2C on I2C bus 1 at 0x1e (bus: 100 KHz, max: 400 KHz)
INFO [lis3mdl] Poll rate set to max (80hz)
INFO [lis3mdl] Set mag range to 4 Gauss
nsh> lis3mdl start
LIS3MDL_SPI on SPI bus 1 at 7 (11000 KHz)
INFO [lis3mdl] Poll rate set to max (80hz)
INFO [lis3mdl] Set mag range to 4 Gauss
nsh> lis3mdl -X status
WARN [lis3mdl] running on bus: 2 (/dev/lis3mdl_ext)

lis3mdl_read: 504 events, 1153644us elapsed, 2288us avg, min 2011us max 3685us 270.915us rms
lis3mdl_comms_errors: 0 events
INFO [lis3mdl] poll interval: 13 ticks
sensor_mag_s
        timestamp: 389383395 (0.008588 seconds ago)
        error_count: 0
        x: 0.799
        y: -0.324
        z: -0.242
        range_ga: 4.000
        scaling: 0.000
        temperature: 35.500
        device_id: 335369
        x_raw: 5471
        y_raw: -2241
        z_raw: -1676
        is_external: 1
report queue        2/112 (0/2 @ 2002ab70)
nsh> lis3mdl -S status
WARN [lis3mdl] running on bus: 3 (/dev/lis3mdl_spi)

lis3mdl_read: 0 events, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
lis3mdl_comms_errors: 0 events
INFO [lis3mdl] poll interval: 0 ticks
sensor_mag_s
        timestamp: 0
        error_count: 0
        x: 0.000
        y: 0.000
        z: 0.000
        range_ga: 0.000
        scaling: 0.000
        temperature: 0.000
        device_id: 0
        x_raw: 0
        y_raw: 0
        z_raw: 0
        is_external: 0
report queue        2/112 (2/2 @ 2002b3b0)
nsh>
```

When I start them the other way around, they run just fine. However, starting order should not matter:

```
nsh> lis3mdl stop
nsh> lis3mdl -S start
LIS3MDL_SPI on SPI bus 1 at 7 (11000 KHz)
INFO [lis3mdl] Poll rate set to max (80hz)
INFO [lis3mdl] Set mag range to 4 Gauss
nsh> lis3mdl -X start
LIS3MDL_I2C on I2C bus 1 at 0x1e (bus: 100 KHz, max: 400 KHz)
INFO [lis3mdl] Poll rate set to max (80hz)
INFO [lis3mdl] Set mag range to 4 Gauss
nsh> lis3mdl -S status
WARN [lis3mdl] running on bus: 3 (/dev/lis3mdl_spi)

lis3mdl_read: 480 events, 413035us elapsed, 860us avg, min 652us max 2378us 170.875us rms
lis3mdl_comms_errors: 0 events
INFO [lis3mdl] poll interval: 13 ticks
sensor_mag_s
        timestamp: 522282524 (0.007740 seconds ago)
        error_count: 0
        x: -0.098
        y: 0.390
        z: 0.987
        range_ga: 4.000
        scaling: 0.000
        temperature: 40.250
        device_id: 329482
        x_raw: -645
        y_raw: 2619
        z_raw: 6761
        is_external: 0
report queue        2/112 (0/2 @ 2002acd0)
nsh> lis3mdl -X status
WARN [lis3mdl] running on bus: 2 (/dev/lis3mdl_ext)

lis3mdl_read: 426 events, 1038242us elapsed, 2437us avg, min 2034us max 4079us 386.869us rms
lis3mdl_comms_errors: 0 events
INFO [lis3mdl] poll interval: 13 ticks
sensor_mag_s
        timestamp: 525748397 (0.008619 seconds ago)
        error_count: 0
        x: 0.801
        y: -0.327
        z: -0.227
        range_ga: 4.000
        scaling: 0.000
        temperature: 35.250
        device_id: 335369
        x_raw: 5512
        y_raw: -2214
        z_raw: -1519
        is_external: 1
report queue        2/112 (0/2 @ 2002b570)
nsh>
```

### PR DESCRIPTION
**Pull request by hdiethelm**
Fixes #10740 

Rebase on master from #10748

lis3mdl was not initializing multiple sensors properly when started using:
lis3mdl start
The second sensor was just initialized but not started and there where some other bugs which where not visible at first glance.

Now, lis3mdl does the actions for all sensors when not called with -I, -X or -S.

I tested all commands with no argument, -X and -S. I was not able to test -I because I don't have an internal I2C sensor.

I tested the code on the stable branch, so it should work also with master:
[test_log.txt](https://github.com/PX4/Firmware/files/2511456/test_log.txt)

I'll do some short tests later.

### CODE REVIEW

**Comment by hdiethelm** [comment]
Just tested and they are working well also in master! i I added an additional message when lis3mdl is started with the -C option and the calibration was successful.

```
nsh> lis3mdl start -C
LIS3MDL_I2C on I2C bus 1 at 0x1e (bus: 100 KHz, max: 400 KHz)
INFO  [lis3mdl] running on bus: 2 (/dev/lis3mdl_ext)
set_excitement enable=1 cntl1=0xfd
INFO  [lis3mdl] calibration successful
INFO  [lis3mdl] Poll rate set to max (80hz)
LIS3MDL_SPI on SPI bus 1 at 7 (11000 KHz)
INFO  [lis3mdl] running on bus: 3 (/dev/lis3mdl_spi)
set_excitement enable=1 cntl1=0xfd
INFO  [lis3mdl] calibration successful
INFO  [lis3mdl] Poll rate set to max (80hz)
```
```
INFO  [sensors] mag status:
INFO  [ecl/validation] validator: best: 0, prev best: 2, failsafe: NO (0 events)
INFO  [ecl/validation] sensor #0, prio: 255, state: OK
INFO  [ecl/validation] 	val:   0.3183, lp:   0.3142 mean dev:  -0.0002 RMS:   0.0030 conf:   1.0000
INFO  [ecl/validation] 	val:   0.0201, lp:   0.0229 mean dev:  -0.0000 RMS:   0.0033 conf:   1.0000
INFO  [ecl/validation] 	val:   0.5557, lp:   0.5634 mean dev:   0.0001 RMS:   0.0049 conf:   1.0000
INFO  [ecl/validation] sensor #1, prio: 100, state: OK
INFO  [ecl/validation] 	val:   0.2713, lp:   0.2738 mean dev:  -0.0000 RMS:   0.0027 conf:   1.0000
INFO  [ecl/validation] 	val:   0.0116, lp:   0.0156 mean dev:   0.0000 RMS:   0.0026 conf:   1.0000
INFO  [ecl/validation] 	val:   0.5337, lp:   0.5391 mean dev:   0.0001 RMS:   0.0066 conf:   1.0000
INFO  [ecl/validation] sensor #2, prio: 50, state: OK
INFO  [ecl/validation] 	val:   0.2780, lp:   0.2726 mean dev:   0.0003 RMS:   0.0073 conf:   1.0000
INFO  [ecl/validation] 	val:   0.0529, lp:   0.0528 mean dev:  -0.0005 RMS:   0.0080 conf:   1.0000
INFO  [ecl/validation] 	val:   0.5432, lp:   0.5426 mean dev:   0.0001 RMS:   0.0087 conf:   1.0000
```

**Comment by LorenzMeier** [comment]
When you say "is working well also in master" what do you mean with this? I don't see that you rebased on master?

**Comment by hdiethelm** [comment]
Sorry, might be I wrote it wrong. What I did:
git checkout master
git submodule update
cherry-pick my commits from my branch
git rebase -i HEAD~3 to squash the tree commits in to one
git push fork
make clean
make px4fmu-v4_default upload
tested lis3mdl
Hope that was all right?

**Comment by LorenzMeier** [comment]
My bad. The commits show up after your first comment, which says it got rebased. It's all good.

File: src/drivers/magnetometer/lis3mdl/lis3mdl_main.cpp [hunk 1]
```diff
@@ -69,55 +70,52 @@ lis3mdl::calibrate(LIS3MDL_BUS bus_id)
...
-lis3mdl::info(LIS3MDL_BUS bus_id)
+lis3mdl::info(struct lis3mdl_bus_option &bus)
 {
-	struct lis3mdl_bus_option &bus = find_bus(bus_id);
-
-	PX4_WARN("running on bus: %u (%s)\n", (unsigned)bus.bus_id, bus.devpath);
+	PX4_INFO("running on bus: %u (%s)", (unsigned)bus.bus_id, bus.devpath);
 	bus.dev->print_info();
-	return 1;
+	return 0;
```
Comment (dakejahl) [inline src/drivers/magnetometer/lis3mdl/lis3mdl_main.cpp]: Use `PX4_OK` and `PX4_ERROR`. This applies for all return values and checks on functions success/fail. This helps keep the logic clean.

File: src/drivers/magnetometer/lis3mdl/lis3mdl_main.cpp [hunk 2]
```diff
@@ -69,55 +70,52 @@ lis3mdl::calibrate(LIS3MDL_BUS bus_id)
...
+lis3mdl::init(struct lis3mdl_bus_option &bus)
 {
-	struct lis3mdl_bus_option &bus = find_bus(bus_id);
 	const char *path = bus.devpath;
 
 	int fd = open(path, O_RDONLY);
 
 	if (fd < 0) {
-		return false;
+		return 1;
```
Comment (dakejahl) [inline src/drivers/magnetometer/lis3mdl/lis3mdl_main.cpp]: Same as above. Apply to all return values.

**Pull request merged**